### PR TITLE
`is_commuting` function accepts opmath instances

### PR DESCRIPTION
**Pull request by PietropaoloFrisoni**
**Context:** We frequently work with Pauli words in the form of opmath instances (`Sum,` `SProd,` and `Prod`) rather than `Tensor.` For this reason, the function `qml.is_commuting` should also work with opmath instances. Such a feature is also necessary to enable the new operation arithmetic by default

**Description of the Change:**  The `qml.is_commuting` function now accepts `Sum,` `SProd,` and `Prod` instances. Since the function currently only accepts `Tensor` instances containing Pauli words, the extension to opmath instances has the same limitation. 

**Benefits:** This change makes it possible to use the `qml.is_commuting` function by providing opmath instances on top of `Tensor` instances.

**Possible Drawbacks:** As described above, the opmath instances must only contain Pauli words. Otherwise, an error is raised. 

**Related GitHub Issues:** None.

[sc-51912]
[sc-58417]

### CODE REVIEW

File: pennylane/ops/functions/is_commuting.py [hunk 1]
```diff
@@ -161,6 +162,36 @@ def commutes_inner(op_name1, op_name2):
 _commutes = _create_commute_function()
 
 
+def _check_opmath_operations(operation1, operation2):
```
Comment (albi3ro) [inline pennylane/ops/functions/is_commuting.py:165]: could potentially  just check `if operation1.pauli_rep is not None`.

Comment (Qottmann) [reply pennylane/ops/functions/is_commuting.py:165]: I agree, this seems like overkill or is there cases that wouldnt be caught with my suggestion [here](https://xanaduhq.slack.com/archives/C06DX049G49/p1709296581611489?thread_ts=1709247345.316049&cid=C06DX049G49)?

Comment (Qottmann) [reply pennylane/ops/functions/is_commuting.py:165]: Aside from the code being simpler it should also be fastest to use the pauli rep of the operators 🙂

Comment (PietropaoloFrisoni) [reply pennylane/ops/functions/is_commuting.py:165]: Thanks both for the suggestion. I didn't know that `.pauli_rep` also works with `sum.` On the other hand, `is_pauli_word` does not recognize sums:

```
p = qml.sum(qml.PauliX(0), qml.prod(qml.PauliY(1), qml.PauliZ(2)))
is_pauli_word(p)
```
results in:

```
False
```

which is why I had to do such a lengthy check. With `.pauli_rep`, the check is much more straightforward.

**Review by albi3ro — COMMENTED** [review]
My general piece of feedback is to rely more on the `pauli_rep` property and `PauliSentence` behaviour.  If it has a `pauli_rep`, we can trust that it is a linear combination of pauli words, independent of its particular format.  We don't need to care whether is `Tensor`, `Sum`, `Prod`, or `SProd`; we can just care whether or not it has a `pauli_rep`.

We can also calculate the commutator using the pauli representation `qml.commutator(op1, op2, pauli=True)`


Note that using pauli reps, we can check the commutation of ops with multiple terms:

```
>>> op1 = qml.X(0) + qml.Y(1)
>>> op2 = qml.Y(1) + qml.Z(2)
>>> ps = qml.commutator(op1, op2, pauli=True)
>>> ps.simplify()
>>> len(ps)
0
```

**Comment by PietropaoloFrisoni** [comment]
I actually like the suggestion that @Qottmann left [here](https://xanaduhq.slack.com/archives/C06DX049G49/p1709296581611489?thread_ts=1709247345.316049&cid=C06DX049G49). But what about the `wire_map` argument that the `is_commuting` function currently accepts? Should we just remove it (in that case, is this a potential breaking change)?

**Comment by PietropaoloFrisoni** [comment]
Maybe we can do something similar in case a `wire_map` is provided:
```
p = qml.prod(qml.PauliZ("a"), qml.PauliY("b"), qml.PauliZ("d"))
WM = {"a": 0, "b": 1, "c": 2, "d": 3}
p = p.map_wires(WM)

p.pauli_rep
```
Which provides:

```
1.0 * Y(1) @ Z(0) @ Z(3)
```

Please tell me if there's a more elegant way to do it (I am still not super familiar with all the nice properties of `PauliSentence` and similar). Otherwise, I'll update the code tomorrow by replacing the `_pword_is_commuting` function

**Comment by PietropaoloFrisoni** [comment]
> If it has a `pauli_rep`, we can trust that it is a linear combination of pauli words, independent of its particular format. We don't need to care whether is `Tensor`, `Sum`, `Prod`, or `SProd`; we can just care whether or not it has a `pauli_rep`.

@albi3ro The purpose of `_check_opmath_operations` is to raise an error if an opmath instance contains an operator that is not a Pauli word. 

As far as I understand, the function should:

- work with single operators that are not Pauli words, such as 
```
qml.is_commuting(qml.X(0), qml.Hadamard(0))
```
- work with `SProd`, `Prod`, `Sum`, and `Tensor` operators that only contain Pauli words.

- raise an error with `SProd`, `Prod`, `Sum`, and `Tensor` operators that do not contain only Pauli words.

Unless I am missing something, it seems impossible to provide a detailed error message without caring whether it is `Tensor,` `Sum,` `Prod,` or `SProd.` If we absolutely don't want to check against all possible cases, should we just raise a generic error message by just checking the `arithmetic_depth` property?

File: pennylane/ops/functions/is_commuting.py [hunk 2]
```diff
@@ -42,24 +36,29 @@ def _pword_is_commuting(pauli_word_1, pauli_word_2, wire_map=None):
...
+    >>> pauli_word_1 = qml.sum(qml.X('a') , qml.Y('b'))
+    >>> pauli_word_2 = qml.sum(qml.Z('c') , qml.X('a'))
+    >>> _pword_is_commuting(pauli_word_1, pauli_word_2, wire_map=wire_map)
+    True
     """
 
     if wire_map is None:
         wire_map = _wire_map_from_pauli_pair(pauli_word_1, pauli_word_2)
+        pauli_word_1 = pauli_word_1.map_wires(wire_map)
+        pauli_word_2 = pauli_word_2.map_wires(wire_map)
```
Comment (albi3ro) [inline pennylane/ops/functions/is_commuting.py:54]: Can we just get rid of this now?

Comment (PietropaoloFrisoni) [reply pennylane/ops/functions/is_commuting.py:54]: So we get rid of the `wire_map` argument that the `is_commuting` function currently accepts?

**Review by albi3ro — APPROVED** [review]

**Pull request merged**